docarray/doc_index/abstract_doc_index.py

```diff
@@ -672,30 +703,48 @@ def _create_single_column(self, field: 'ModelField', type_: Type) -> _ColumnInfo
             docarray_type=type_, db_type=db_type, config=config, n_dim=n_dim
         )
 
-    def _is_schema_compatible(self, docs: Sequence[BaseDocument]) -> bool:
-        """Flatten a DocumentArray into a DocumentArray of the schema type."""
-        reference_col_db_types = [
-            (name, col.db_type) for name, col in self._column_infos.items()
-        ]
-        if isinstance(docs, AnyDocumentArray):
-            input_columns = self._create_columns(docs.document_type)
-            input_col_db_types = [
-                (name, col.db_type) for name, col in input_columns.items()
-            ]
+    def _validate_docs(
+        self, docs: Union[BaseDocument, Sequence[BaseDocument]]
+    ) -> DocumentArray[BaseDocument]:
+        if isinstance(docs, BaseDocument):
+            docs = [docs]
+        if isinstance(docs, DocumentArray):
+            # validation shortcut for DocumentArray; only look at the schema
+            reference_schema_flat = self._flatten_schema(
+                cast(Type[BaseDocument], self._schema)
+            )
+            reference_names = [name for (name, _, _) in reference_schema_flat]
+            reference_types = [t_ for (_, t_, _) in reference_schema_flat]
+
+            input_schema_flat = self._flatten_schema(docs.document_type)
+            input_names = [name for (name, _, _) in input_schema_flat]
+            input_types = [t_ for (_, t_, _) in input_schema_flat]
             # this could be relaxed in the future,
             # see schema translation ideas in the design doc
-            return reference_col_db_types == input_col_db_types
-        else:
-            for d in docs:
-                input_columns = self._create_columns(type(d))
-                input_col_db_types = [
-                    (name, col.db_type) for name, col in input_columns.items()
-                ]
-                # this could be relaxed in the future,
-                # see schema translation ideas in the design doc
-                if reference_col_db_types != input_col_db_types:
-                    return False
-            return True
+            names_compatible = reference_names == input_names
+            types_compatible = all(
+                (not is_union_type(t2) and issubclass(t1, t2))
+                for (t1, t2) in zip(reference_types, input_types)
+            )
+            if names_compatible and types_compatible:
```

Comment:
so if this is False, docs get checked again and only after the error is thrown, right? should we somehow separate these cases so that we don't have to check docs twice? especially since the second check is a costly operation as we do it for each doc

Reply:
> Only if this is not successful we iterate over the docs

Do we need to iterate over docs if we already know that it's not successful? this was my point

But I guess if the schema is not compatible, error will be thrown right away for the first doc at line 767 so it's not really an iteration over all docs

Reply:
yes, we still need to check because the actual data might be compatible even if the schema says otherwise.

For example, the input could be of type `DocumentArray[BaseDocument]`, but contain only `ImageDoc`. This is valid since `ImageDoc` is a subclass of `BaseDocument`.
If the Document Index has schema `ImageDoc`, then the fast validation will fail, because `BaseDoc != ImageDoc`, but actually it should work, since all Documents are actually `ImageDoc`. So we have to do the exhaustive check over the data.

So basically the validation check based on the DocumentArray parametrization can be seen as just a shortcut that may or may not work.

Reply:
makes sense, thanks for clarification!